File: 2023/8.py
from math import lcm

input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n\n")
instructions, nodes_input = input

nodes = {}
for n in nodes_input.split("\n"):
    s = n.split(" = ")
    nodes[s[0]] = s[1][1:-1].split(", ")


steps = 0
current_node = "AAA"
while True:
    if current_node == "ZZZ":
        print(steps)
        break

    current_node = (
        nodes[current_node][0]
        if instructions[steps % len(instructions)] == "L"
        else nodes[current_node][1]
    )
    steps += 1

current_nodes = list(filter(lambda x: x[-1] == "A", nodes.keys()))
loops = []
for current_node in current_nodes:
    steps = 0
    while True:
        if current_node[-1] == "Z":
            loops.append(steps)
            break

        current_node = (
            nodes[current_node][0]
            if instructions[steps % len(instructions)] == "L"
            else nodes[current_node][1]
        )
        steps += 1

print(lcm(*loops))

File: 2023/8.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input'))
instructions, nodes = input.split("\n\n")

nodes = nodes.split("\n").map { _1.split(' = ') }.map { [_1[0], _1[1][1..-2].split(', ')] }.to_h

steps = 0
current_node = 'AAA'
loop do
  if current_node == 'ZZZ'
    p steps
    break
  end

  current_node = instructions[steps % instructions.size] == 'L' ? nodes[current_node][0] : nodes[current_node][1]
  steps += 1
end

loops = []
nodes.keys.select { _1[-1] == 'A' }.map do |current_node|
  steps = 0
  loop do
    if current_node[-1] == 'Z'
      loops << steps
      break
    end

    current_node = instructions[steps % instructions.size] == 'L' ? nodes[current_node][0] : nodes[current_node][1]
    steps += 1
  end
end

p loops.reduce(1, :lcm)
